concordium-consensus/src/Concordium/GlobalState/AccountMap/DifferenceMap.hs

```diff
@@ -0,0 +1,139 @@
+{-# LANGUAGE BangPatterns #-}
+
+-- | The 'DifferenceMap' stores accounts that have been created in a non-finalized block.
+--  When a block is finalized then the associated 'DifferenceMap' must be written
+--  to disk via 'Concordium.GlobalState.AccountMap.LMDB.insertAccounts'.
+module Concordium.GlobalState.AccountMap.DifferenceMap (
+    -- * Definitions
+
+    -- The difference map definition.
+    DifferenceMap (..),
+    -- A mutable reference to a 'DifferenceMap'.
+    DifferenceMapReference,
+
+    -- * Auxiliary functions
+
+    -- The empty reference
+    emptyReference,
+    -- Get a list of all @(AccountAddress, AccountIndex)@ pairs for the
+    --  provided 'DifferenceMap' and all parent maps.
+    flatten,
+    -- Create an empty 'DifferenceMap'
+    empty,
+    -- Set the accounts int he 'DifferenceMap'.
+    fromList,
+    -- Insert an account into the 'DifferenceMap'.
+    insert,
+    -- Lookup in a difference map (and potential parent maps) whether
+    -- it yields the 'AccountIndex' for the provided 'AccountAddress' or any
+    -- alias of it.
+    lookupViaEquivalenceClass,
+    -- Lookup in a difference map (and potential parent maps) whether
+    -- it yields the 'AccountIndex' for the provided 'AccountAddress'.
+    lookupExact,
+) where
+
+import Control.Monad.IO.Class
+import Data.Bifunctor
+import Data.Foldable
+import qualified Data.HashMap.Strict as HM
+import Data.IORef
+import Prelude hiding (lookup)
+
+import Concordium.Types
+import Concordium.Types.Option (Option (..))
+
+-- | A mutable reference to a 'DiffMap.DifferenceMap'.
+type DifferenceMapReference = IORef (Option DifferenceMap)
+
+-- | The empty reference
+emptyReference :: (MonadIO m) => m (IORef (Option DifferenceMap))
+emptyReference = liftIO $ newIORef Absent
```

Comment:
I think it might be worth calling this `newEmptyReference`. There could also be a `finalizedEmptyReference`, which is `unsafePerformIO emptyReference` (with a no-inline pragma). This would serve as a common empty reference that can be used when it is guaranteed that the contents of the reference will never be changed.

Reply:
I've renamed to `newEmptyReference`. Not sure if it's worth it to add the `finalizedEmptyReference`, there will only ever be 1 alive right? So in practice there will be two empty references (as opposed to 1 qua your suggestion) will be from a (child to a block A) is finalized and until A is GC'ed. @td202 @abizjak  This seems like a premature optimization. But it can be that I'm wrong and in that case I am happy to revise this. 

Resolving for now.